src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -1105,14 +1108,15 @@ private void noteReceived(MidiNote note, boolean on) {
           } else {
             LXClip clip = channel.getClip(index);
             if (clip == null) {
-              clip = channel.addClip(index);
+              channel.addClip(index, this.shiftOn);
```

Comment:
Change this to `clip = channel.addClip(index)` followed by:
```
if (this.shiftOn) {
  clip.loop.setValue(true);
}
```

Then pop over to lines 548 - 559, and add:
```
clip.loop.addListener(this);
clip.loop.removeListener(this);
```

Adjacent to where clip.running is monitored. This will ensure the right thing happens whenever and however the looping status of the clip is modified, whether by control surface or otherwise.

Also need to do that on lines 437/461 in the ChannelListener contructor and dispose method.

Reply:
Done, though I inlined line 1116 to avoid another layer of braces.